From cf1ad072163003e5df2aa9e328e02f9b3a778fd6 Mon Sep 17 00:00:00 2001
From: Christian Seiler <christian@iwakd.de>
Date: Thu, 3 Mar 2016 01:17:21 +0100
Subject: [PATCH 1/2] adt-virt-qemu: Implement support for nested base images.

This adds initial support for nested base images to be passed into the
test environment, so that nested VMs may be used in tests.

A read-only copy of the first image without the overlay is passed to
the VM with a hardware serial number BASEIMAGE. setup-testbed installs
udev rules that create a symbolic link /dev/baseimage for that drive.
Also, the symlink priority for that drive is lowered, because the same
file system UUIDs will be present on both the first drive and the
readonly baseimage drive.

Closes: #800845
---
 debian/changelog             |  5 +++++
 setup-commands/setup-testbed | 11 +++++++++++
 virt-subproc/adt-virt-qemu   |  2 ++
 virt-subproc/adt-virt-qemu.1 | 20 ++++++++++++++++++++
 4 files changed, 38 insertions(+)

diff --git a/debian/changelog b/debian/changelog
index 36ee620..6a3107b 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,5 +1,6 @@
 autopkgtest (3.19.4) UNRELEASED; urgency=medium
 
+  [ Martin Pitt ]
   * setup-commands/setup-testbed: Ensure that removing cruft does not remove
     cloud-init. (LP: #1539126)
   * setup-commands/setup-testbed: Purge lxd and lxc.
@@ -14,6 +15,10 @@ autopkgtest (3.19.4) UNRELEASED; urgency=medium
     lxc-start-ephemeral got deprecated by that. This now supports reboots in
     ephemeral mode.
 
+  [ Christian Seiler ]
+  * adt-virt-qemu: Implement support for nested base images.
+    (Closes: #800845)
+
  -- Martin Pitt <mpitt@debian.org>  Tue, 23 Feb 2016 18:21:51 +0100
 
 autopkgtest (3.19.3) unstable; urgency=medium
diff --git a/setup-commands/setup-testbed b/setup-commands/setup-testbed
index b7f586f..bcdbad3 100755
--- a/setup-commands/setup-testbed
+++ b/setup-commands/setup-testbed
@@ -142,6 +142,17 @@ if [ -z "${ADT_IS_SETUP_COMMAND:-}" ]; then
     fi
 fi
 
+# set up base image udev rules
+if [ -z "${ADT_IS_SETUP_COMMAND:-}" ]; then
+    # Create /dev/baseimage symlink for the base image, and
+    # make sure that the link_priority is low, so that the
+    # drive has a lower priority for udev symlinks. (Because
+    # the same base image as for the main drive is reused.)
+    printf '%s\n%s\n' 'KERNEL=="vd*[!0-9]", ENV{ID_SERIAL}=="BASEIMAGE", OPTIONS+="link_priority=-1024", SYMLINK+="baseimage", MODE="0664"' \
+                      'KERNEL=="vd*[0-9]",  ENV{ID_SERIAL}=="BASEIMAGE", OPTIONS+="link_priority=-1024"' \
+                     > "$root/etc/udev/rules.d/61-baseimage.rules"
+fi
+
 # go-faster apt/dpkg
 echo "Acquire::Languages \"none\";" > "$root"/etc/apt/apt.conf.d/90nolanguages
 echo 'force-unsafe-io' > "$root"/etc/dpkg/dpkg.cfg.d/autopkgtest
diff --git a/virt-subproc/adt-virt-qemu b/virt-subproc/adt-virt-qemu
index 965e3e8..d9ae8c8 100755
--- a/virt-subproc/adt-virt-qemu
+++ b/virt-subproc/adt-virt-qemu
@@ -476,6 +476,8 @@ def hook_open():
     for i, image in enumerate(args.image[1:]):
         argv.append('-drive')
         argv.append('file=%s,if=virtio,index=%i,readonly' % (image, i + 1))
+    argv.append('-drive')
+    argv.append('file=%s,if=virtio,index=%i,readonly,serial=BASEIMAGE' % (args.image[0], len(args.image)))
 
     if os.path.exists('/dev/kvm'):
         argv.append('-enable-kvm')
diff --git a/virt-subproc/adt-virt-qemu.1 b/virt-subproc/adt-virt-qemu.1
index 81bb11f..ed888e6 100644
--- a/virt-subproc/adt-virt-qemu.1
+++ b/virt-subproc/adt-virt-qemu.1
@@ -26,6 +26,12 @@ does
 the given images, but will instead create a temporary overlay for the
 primary image, and add all other images as read-only.
 
+The first image without the overlay is always added as an additional
+read-only hard drive, which will be available for tests as
+.IR /dev/baseimage ,
+assuming the correct udev rules are present within the image. This allows
+for tests that require nested VMs to reuse the same image.
+
 .SH REQUIREMENTS
 .B adt-virt-qemu
 assumes that you have already prepared a suitable Debian based QEMU image (see
@@ -45,6 +51,20 @@ with specified user and password. This will then be used to start a root
 shell on ttyS1, to reduce this to the first case and to not assume that
 ttyS0 stays operational throughout the whole test.
 
+.IP * 4
+and has the following udev rule installed as
+.IR /etc/udev/rules.d/61-baseimage.rules :
+
+.RS
+.EX
+KERNEL=="vd*[!0-9]", ENV{ID_SERIAL}=="BASEIMAGE", \\
+  OPTIONS+="link_priority=-1024", SYMLINK+="baseimage", \\
+  MODE="0664"
+KERNEL=="vd*[0-9]",  ENV{ID_SERIAL}=="BASEIMAGE", \\
+  OPTIONS+="link_priority=-1024"
+.EE
+.RE
+
 .SH OPTIONS
 
 .TP
-- 
2.1.4

